dagger/input/validators.py

```diff
@@ -26,29 +24,62 @@ def validate_name(name: str):
         )
 
 
-def validate_parameters(
+def split_required_and_optional_inputs(
     inputs: Mapping[str, Union[FromParam, FromNodeOutput]],
-    params: Mapping[str, Any],
-):
+) -> Tuple[Mapping[str, Union[FromParam, FromNodeOutput]], Mapping[str, FromParam],]:
     """
-    Validate a series of parameters against the inputs of a DAG.
+    Split a map of inputs into a tuple of (required, optional) input maps.
 
     Parameters
     ----------
     inputs
         A mapping of input names to inputs.
 
     params
-        A mapping of input names to parameters or input values.
-        Input values must be passed in their serialized representation.
+        A mapping of input names to their values.
+
+
+    Returns
+    -------
+    A 2-tuple of (required, optional) input mappings.
+    """
+    required, optional = {}, {}
+
+    for input_name, input_type in inputs.items():
+        if isinstance(input_type, FromParam) and input_type.has_default_value():
+            optional[input_name] = input_type
+        else:
+            required[input_name] = input_type
+
+    return required, optional
+
+
+def validate_and_clean_parameters(
+    inputs: Mapping[str, Union[FromParam, FromNodeOutput]], params: Mapping[str, Any]
+) -> Mapping[str, Any]:
```

Comment:
Yeah, I had the same doubts as I was writing this piece of code.

If we went for 2 separate functions `validate()` + `clean()`, we would have N places in the code where we'd need to use:

```python
validate(inputs, params)
params = clean(inputs, params)
```

The clean function would have as a precondition the validity of the parameters, so those 2 functions will always need to be called in tandem. Thus, it would feel strange to provide them separately. That would give the users of this module the impression that they are independent.

It is definitely awkward to see the name `x_and_y()`, but maybe in this case it's just a problem with the name.

In any case, I'm happy to change it if you feel it'd be better

Reply:
I liked you explanation. What do you think about creating two private functions, `_validate_parameters` and `_clean_parameters` that are called by `validate_and_clean_parameters`?

Reply:
I like that idea. Let's do it!